File: backend/schemas/knowledge.py
"""
知识库 + 文档 请求/响应 Schema

- KnowledgeBaseCreate / Update：创建/编辑请求
- KnowledgeBaseResponse：详情 + 创建/编辑 响应（完整字段）
- KnowledgeBaseListItem：列表响应（精简，仅表格需要）
- DocumentResponse：上传响应（完整）
- DocumentListItem：文件列表响应（精简）
"""
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from backend.schemas.cleaning import CleaningConfig


# ── 知识库 ────────────────────────────

class KnowledgeBaseCreate(BaseModel):
    """创建知识库"""
    name: str = Field(..., min_length=1, max_length=200)
    description: str | None = Field(None)
    allowed_extensions: str | None = Field(None, max_length=500)
    max_file_size: int | None = Field(None, ge=1, description="单文件上限（字节），不传则使用 config.yaml 默认值")
    allow_multiple: bool = Field(True)
    auto_process: bool | None = Field(None, description="上传后自动处理；不传则复制当前系统的新建默认值")
    chunk_size: int | None = Field(None, ge=50, le=10000)
    chunk_overlap: int | None = Field(None, ge=0, le=9999)
    chunk_separators: str | None = None
    cleaning_config: CleaningConfig | None = None
    embedding_model: str | None = None
    embedding_dimension: int | None = None
    user_model_config_id: UUID | None = Field(None, description="关联的用户模型配置，NULL 则使用本地默认 Embedding 模型")

    @model_validator(mode="after")
    def validate_chunk_config(self) -> "KnowledgeBaseCreate":
        """确保重叠长度小于单片长度。"""
        if (
            self.chunk_size is not None
            and self.chunk_overlap is not None
            and self.chunk_overlap >= self.chunk_size
        ):
            raise ValueError("chunk_overlap 必须小于 chunk_size")
        return self


class KnowledgeBaseUpdate(BaseModel):
    """编辑知识库（全部可选）"""
    name: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = None
    allowed_extensions: str | None = Field(None, max_length=500)
    max_file_size: int | None = Field(None, ge=1)
    allow_multiple: bool | None = None
    auto_process: bool | None = None
    chunk_size: int | None = Field(None, ge=50, le=10000)
    chunk_overlap: int | None = Field(None, ge=0, le=9999)
    chunk_separators: str | None = None
    cleaning_config: CleaningConfig | None = None
    embedding_model: str | None = None
    embedding_dimension: int | None = None
    user_model_config_id: UUID | None = None
    status: int | None = Field(None, ge=0, le=9)

    @model_validator(mode="after")
    def validate_chunk_config(self) -> "KnowledgeBaseUpdate":
        """当两个切片参数同时提交时，确保重叠长度小于单片长度。"""
        if (
            self.chunk_size is not None
            and self.chunk_overlap is not None
            and self.chunk_overlap >= self.chunk_size
        ):
            raise ValueError("chunk_overlap 必须小于 chunk_size")
        return self


class KnowledgeBaseResponse(BaseModel):
    """知识库详情/编辑 响应（编辑弹窗回填需要）"""
    id: UUID
    name: str
    description: str | None
    allowed_extensions: str | None
    max_file_size: int
    allow_multiple: bool
    auto_process: bool
    chunk_size: int
    chunk_overlap: int
    chunk_separators: str
    cleaning_config: dict
    embedding_model: str
    embedding_dimension: int
    user_model_config_id: UUID | None
    status: int
    created_at: datetime

    model_config = {"from_attributes": True}


class KnowledgeBaseListItem(BaseModel):
    """知识库列表项（仅表格需要）"""
    id: UUID
    name: str
    allowed_extensions: str | None
    user_model_config_id: UUID | None = None
    embedding_model: str = ""
    status: int
    created_at: datetime

    model_config = {"from_attributes": True}


# ── 文档 ──────────────────────────────

class ProcessingConfig(BaseModel):
    """可用于知识库或单个文件的完整清洗与切片策略。"""

    cleaning_config: CleaningConfig
    chunk_size: int = Field(..., ge=50, le=10000)
    chunk_overlap: int = Field(..., ge=0, le=9999)
    chunk_separators: str

    @model_validator(mode="after")
    def validate_chunk_config(self) -> "ProcessingConfig":
        """确保重叠长度小于单片长度。"""
        if self.chunk_overlap >= self.chunk_size:
            raise ValueError("chunk_overlap 必须小于 chunk_size")
        return self


class DocumentProcessingStrategyUpdate(BaseModel):
    """切换文件策略继承模式，或保存一份完整的文件独立策略。"""

    mode: Literal["inherit", "custom"]
    cleaning_config: CleaningConfig | None = None
    chunk_size: int | None = Field(None, ge=50, le=10000)
    chunk_overlap: int | None = Field(None, ge=0, le=9999)
    chunk_separators: str | None = None

    @model_validator(mode="after")
    def validate_mode_payload(self) -> "DocumentProcessingStrategyUpdate":
        """inherit 不接受独立字段；custom 必须提供完整且有效的策略。"""
        fields = (
            self.cleaning_config,
            self.chunk_size,
            self.chunk_overlap,
            self.chunk_separators,
        )
        if self.mode == "inherit":
            if any(value is not None for value in fields):
                raise ValueError("inherit 模式不能携带 custom 策略字段")
            return self
        if any(value is None for value in fields):
            raise ValueError("custom 模式必须提供完整处理策略")
        ProcessingConfig(
            cleaning_config=self.cleaning_config,
            chunk_size=self.chunk_size,
            chunk_overlap=self.chunk_overlap,
            chunk_separators=self.chunk_separators,
        )
        return self

    def to_processing_config(self) -> ProcessingConfig | None:
        """将 custom 请求转换为规范策略；inherit 返回 None。"""
        if self.mode == "inherit":
            return None
        return ProcessingConfig(
            cleaning_config=self.cleaning_config,
            chunk_size=self.chunk_size,
            chunk_overlap=self.chunk_overlap,
            chunk_separators=self.chunk_separators,
        )

class PreviewChunk(BaseModel):
    """仅用于前端预览的内存切片。"""

    index: int
    content: str
    char_count: int
    overlap_char_count: int = 0


class DocumentPreview(BaseModel):
    """未落库的文档清洗与切片预览。"""

    document_id: UUID
    filename: str
    raw_content: str
    cleaned_content: str
    chunks: list[PreviewChunk]
    cleaning_config_hash: str
    warnings: list[str] = Field(default_factory=list)


class DocumentPreviewSource(BaseModel):
    """供浏览器本地清洗和切片的原始解析文本。"""

    document_id: UUID
    filename: str
    raw_content: str
    algorithm_version: int
    truncated: bool = False
    warnings: list[str] = Field(default_factory=list)


class StoredChunkItem(BaseModel):
    """数据库当前生效切片的安全展示结构，不包含向量。"""

    id: UUID
    index: int
    content: str
    char_count: int
    metadata: dict | None = None
    is_vectorized: int


class DocumentResponse(BaseModel):
    """文档详情/上传 响应（完整字段）"""
    id: UUID
    knowledge_base_id: UUID
    user_id: UUID
    filename: str
    file_path: str
    file_size: int
    file_type: str | None
    status: int
    is_parsed: int
    is_cleaned: int
    is_chunked: int
    is_vectorized: int
    cleaned_content: str | None = None
    cleaning_config_hash: str | None = None
    last_error: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DocumentListItem(BaseModel):
    """文件列表项（仅表格需要）"""
    id: UUID
    filename: str
    file_size: int
    file_type: str | None
    status: int
    is_parsed: int
    is_cleaned: int
    is_chunked: int
    is_vectorized: int
    processing_config: ProcessingConfig | None = None
    strategy_mode: Literal["inherit", "custom"] = "inherit"
    effective_processing_config: ProcessingConfig | None = None
    strategy_outdated: bool = False
    applied_algorithm_version: int | None = None
    chunk_count: int = 0
    last_error: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── 批量操作 ──────────────────────────────

class BatchKnowledgeBaseAction(BaseModel):
    """批量操作知识库"""
    ids: list[UUID] = Field(..., min_length=1, max_length=100)
    action: Literal["enable", "disable", "delete"]


class BatchFileAction(BaseModel):
    """批量操作文件"""
    ids: list[UUID] = Field(..., min_length=1, max_length=100)
    action: Literal["delete"]


class DocIdsRequest(BaseModel):
    """文档处理请求（解析/分块共用）"""
    doc_ids: list[UUID] = Field(..., min_length=1, max_length=100)


class ProcessDocumentsRequest(DocIdsRequest):
    """将文档处理到指定阶段；服务端自动补齐前置阶段。"""

    target_stage: Literal["parse", "clean", "chunk", "vectorize"] = "vectorize"
    algorithm_version: int = Field(..., ge=1)


class ChunkIdsRequest(BaseModel):
    """向量化请求"""
    chunk_ids: list[UUID] = Field(..., min_length=1, max_length=500)


class ProcessResult(BaseModel):
    """处理结果"""
    processed: int
    total: int


class BatchResult(BaseModel):
    """批量操作结果"""
    success_count: int
    fail_count: int
    failed_ids: list[UUID] = []


# ── 分页 ──────────────────────────────

class PaginatedResponse(BaseModel):
    """分页响应"""
    items: list
    total: int
    page: int
    page_size: int

File: frontend/src/api/knowledge.ts
/**
 * 知识库 + 文件管理 API
 */
import request from './request'

// ── 类型定义 ────────────────────────────

export type CleaningRuleType =
  | 'normalize_unicode'
  | 'remove_control_chars'
  | 'collapse_blank_lines'
  | 'trim_lines'
  | 'collapse_punctuation'
  | 'replace_text'
  | 'regex_remove'

export interface CleaningRule {
  type: CleaningRuleType
  enabled?: boolean
  value?: string | null
  replacement?: string
  max_consecutive?: number
}

export interface CleaningConfig {
  rules: CleaningRule[]
}

export interface ProcessingConfig {
  cleaning_config: CleaningConfig
  chunk_size: number
  chunk_overlap: number
  chunk_separators: string
}

export interface KnowledgeBaseListItem {
  id: string
  name: string
  allowed_extensions: string | null
  user_model_config_id: string | null
  embedding_model: string
  status: number
  created_at: string
}

export interface KnowledgeBaseDetail {
  id: string
  name: string
  description: string | null
  allowed_extensions: string | null
  max_file_size: number
  allow_multiple: boolean
  auto_process: boolean
  chunk_size: number
  chunk_overlap: number
  chunk_separators: string
  cleaning_config: CleaningConfig
  embedding_model: string
  embedding_dimension: number
  user_model_config_id: string | null
  status: number
  created_at: string
}

export interface KnowledgeBaseCreate {
  name: string
  description?: string | null
  allowed_extensions?: string | null
  max_file_size?: number | null
  allow_multiple?: boolean
  auto_process?: boolean | null
  chunk_size?: number | null
  chunk_overlap?: number | null
  chunk_separators?: string | null
  cleaning_config?: CleaningConfig | null
  embedding_model?: string | null
  embedding_dimension?: number | null
  user_model_config_id?: string | null
}

export interface KnowledgeBaseUpdate {
  name?: string | null
  description?: string | null
  allowed_extensions?: string | null
  max_file_size?: number | null
  allow_multiple?: boolean | null
  auto_process?: boolean | null
  chunk_size?: number | null
  chunk_overlap?: number | null
  chunk_separators?: string | null
  cleaning_config?: CleaningConfig | null
  embedding_model?: string | null
  embedding_dimension?: number | null
  user_model_config_id?: string | null
  status?: number | null
}

export interface DocumentListItem {
  id: string
  filename: string
  file_size: number
  file_type: string | null
  status: number
  is_parsed: number
  is_cleaned: number
  is_chunked: number
  is_vectorized: number
  processing_config?: ProcessingConfig | null
  strategy_mode: 'inherit' | 'custom'
  effective_processing_config: ProcessingConfig
  strategy_outdated: boolean
  applied_algorithm_version?: number | null
  chunk_count: number
  last_error?: string | null
  created_at: string
}

export interface DocumentDetail {
  id: string
  knowledge_base_id: string
  user_id: string
  filename: string
  file_path: string
  file_size: number
  file_type: string | null
  status: number
  is_parsed: number
  is_cleaned: number
  is_chunked: number
  is_vectorized: number
  cleaned_content?: string | null
  cleaning_config_hash?: string | null
  last_error?: string | null
  created_at: string
  updated_at: string
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
}

// ── 知识库 CRUD ────────────────────────

/** 获取知识库列表 */
export function listKnowledgeBases(page = 1, pageSize = 20): Promise<PaginatedResponse<KnowledgeBaseListItem>> {
  return request.get('/v1/knowledge-bases', {
    params: { page: String(page), page_size: String(pageSize) },
  })
}

/** 获取知识库详情 */
export function getKnowledgeBase(id: string): Promise<KnowledgeBaseDetail> {
  return request.get(`/v1/knowledge-bases/${id}`)
}

/** 获取新建知识库时使用的服务端默认处理策略。 */
export function getKnowledgeBaseDefaults(): Promise<{
  cleaning_config: CleaningConfig
  auto_process?: boolean
  chunk_size?: number
  chunk_overlap?: number
  chunk_separators?: string
}> {
  return request.get('/v1/knowledge-bases/defaults')
}

/** 创建知识库 */
export function createKnowledgeBase(data: KnowledgeBaseCreate): Promise<KnowledgeBaseDetail> {
  return request.post('/v1/knowledge-bases', data)
}

/** 编辑知识库 */
export function updateKnowledgeBase(id: string, data: KnowledgeBaseUpdate): Promise<KnowledgeBaseDetail> {
  return request.put(`/v1/knowledge-bases/${id}`, data)
}

/** 删除知识库 */
export function deleteKnowledgeBase(id: string): Promise<void> {
  return request.delete(`/v1/knowledge-bases/${id}`)
}

// ── 文件管理 ────────────────────────────

/** 获取文件列表 */
export function listFiles(kbId: string, page = 1, pageSize = 20): Promise<PaginatedResponse<DocumentListItem>> {
  return request.get(`/v1/knowledge-bases/${kbId}/files`, {
    params: { page: String(page), page_size: String(pageSize) },
  })
}

/** 上传文件 */
export function uploadFile(kbId: string, file: File): Promise<DocumentDetail> {
  const formData = new FormData()
  formData.append('file', file)
  return request.post(`/v1/knowledge-bases/${kbId}/files`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

/** 删除文件 */
export function deleteFile(kbId: string, docId: string): Promise<void> {
  return request.delete(`/v1/knowledge-bases/${kbId}/files/${docId}`)
}

export interface PreviewChunk {
  index: number
  content: string
  char_count: number
  overlap_char_count: number
}

export interface DocumentPreview {
  document_id: string
  filename: string
  raw_content: string
  cleaned_content: string
  chunks: PreviewChunk[]
  cleaning_config_hash: string
  warnings: string[]
}

export interface DocumentPreviewSource {
  document_id: string
  filename: string
  raw_content: string
  algorithm_version: number
  truncated: boolean
  warnings: string[]
}

export interface StoredChunk {
  id: string
  index: number
  content: string
  char_count: number
  metadata?: Record<string, unknown> | null
  is_vectorized: number
}

export type DocumentStrategyPayload =
  | { mode: 'inherit' }
  | ({ mode: 'custom' } & ProcessingConfig)

/** 在内存中预览文件的清洗文本和切片，不写入知识库 */
export function previewDocument(kbId: string, docId: string): Promise<DocumentPreview> {
  return request.post(`/v1/knowledge-bases/${kbId}/files/${docId}/preview`)
}

/** 获取一次原始解析文本，供浏览器本地实时清洗和切片。 */
export function getDocumentPreviewSource(kbId: string, docId: string): Promise<DocumentPreviewSource> {
  return request.get(`/v1/knowledge-bases/${kbId}/files/${docId}/preview-source`)
}

/** 分页获取数据库当前生效切片。 */
export function listDocumentChunks(
  kbId: string,
  docId: string,
  page = 1,
  pageSize = 20,
): Promise<PaginatedResponse<StoredChunk>> {
  return request.get(`/v1/knowledge-bases/${kbId}/files/${docId}/chunks`, {
    params: { page: String(page), page_size: String(pageSize) },
  })
}

/** 保存文件独立处理策略，或恢复继承知识库默认策略。 */
export function updateDocumentProcessingStrategy(
  kbId: string,
  docId: string,
  payload: DocumentStrategyPayload,
): Promise<DocumentListItem> {
  return request.put(`/v1/knowledge-bases/${kbId}/files/${docId}/processing-strategy`, payload)
}

// ── 批量操作 ────────────────────────────

export type BatchAction = 'enable' | 'disable' | 'delete'
export type BatchFileAction = 'delete'

export interface BatchResult {
  success_count: number
  fail_count: number
  failed_ids: string[]
}

/** 批量操作知识库 */
export function batchKnowledgeBases(ids: string[], action: BatchAction): Promise<BatchResult> {
  return request.post('/v1/knowledge-bases/batch', { ids, action })
}

/** 批量操作文件 */
export function batchFiles(kbId: string, ids: string[], action: BatchFileAction): Promise<BatchResult> {
  return request.post(`/v1/knowledge-bases/${kbId}/files/batch`, { ids, action })
}

// ── 文档处理 ────────────────────────────

export interface ProcessResult {
  processed: number
  total: number
}

/** 触发文档解析 */
export function parseDocuments(kbId: string, docIds: string[]): Promise<ProcessResult> {
  return request.post(`/v1/knowledge-bases/${kbId}/parse`, { doc_ids: docIds })
}

/** 触发文档分块 */
export function chunkDocuments(kbId: string, docIds: string[]): Promise<ProcessResult> {
  return request.post(`/v1/knowledge-bases/${kbId}/chunk`, { doc_ids: docIds })
}

/** 触发向量化（按 chunk_ids） */
export function vectorizeChunks(kbId: string, chunkIds: string[]): Promise<ProcessResult> {
  return request.post(`/v1/knowledge-bases/${kbId}/vectorize`, { chunk_ids: chunkIds })
}

/** 按文档批量向量化（选文档自动找其所有有效分块） */
export function vectorizeDocuments(kbId: string, docIds: string[]): Promise<ProcessResult> {
  return request.post(`/v1/knowledge-bases/${kbId}/vectorize-docs`, { doc_ids: docIds })
}

export type ProcessTarget = 'parse' | 'clean' | 'chunk' | 'vectorize'

/** 将文件入队处理至指定阶段；后端自动补齐前置阶段。 */
export function processDocuments(
  kbId: string,
  docIds: string[],
  algorithmVersion: number,
  targetStage: ProcessTarget = 'vectorize',
): Promise<ProcessResult> {
  return request.post(`/v1/knowledge-bases/${kbId}/files/process`, {
    doc_ids: docIds,
    target_stage: targetStage,
    algorithm_version: algorithmVersion,
  })
}

// ── 全局处理任务（处理浮窗轮询） ────────

export interface ProcessingTask {
  doc_id: string
  filename: string
  kb_id: string
  kb_name: string
  is_parsed: number
  is_cleaned: number
  is_chunked: number
  is_vectorized: number
  stage: 'queued' | 'parsing' | 'cleaning' | 'chunking' | 'vectorizing' | 'processing'
}

/** 全局处理中/排队任务列表（所有知识库） */
export function getProcessingTasks(): Promise<ProcessingTask[]> {
  return request.get('/v1/knowledge-bases/processing-tasks')
}
